public/js/src/module/connection.js

```diff
@@ -404,6 +396,70 @@ const getTransformURL = (basePath, enketoId) => {
     return `${basePath}/transform/xform${idPath}${_getQuery()}`;
 };
 
+/**
+ * @typedef GetExternalDataOptions
+ * @property {boolean} [isPreview]
+ */
+
+/**
+ * @param {Survey} survey
+ * @param {Document} model
+ * @param {GetExternalDataOptions} [options]
+ * @return {Promise<SurveyExternalData[]>}
+ */
+const getExternalData = async (survey, model, options = {}) => {
+    /** @type {Array<Promise<SurveyExternalData>>} */
+    const tasks = [];
+    const externalInstances = [
+        ...model.querySelectorAll('instance[id][src]'),
+    ].map((instance) => ({
+        id: instance.id,
+        src: instance.getAttribute('src'),
+    }));
+
+    externalInstances.forEach((instance, index) => {
+        const { src } = instance;
+
+        if (src === LAST_SAVED_VIRTUAL_ENDPOINT) {
+            tasks.push(Promise.resolve(instance));
+
+            return;
+        }
+
+        const task = async () => {
+            try {
+                const xml = await getDataFile(src, survey.languageMap);
+
+                return {
+                    ...instance,
+                    xml,
+                };
+            } catch (error) {
+                tasks.splice(index, 1);
+
+                if (options.isPreview) {
```

Comment:
This is the substantive change here, it unifies the two preview modes. Passed through from `getFormParts`